app/services/cart_items_management_service.rb

```diff
@@ -0,0 +1,57 @@
+class CartItemsManagementService
+  attr_reader :cart, :product, :quantity
+
+  def initialize(cart, product, quantity)
+    @cart = cart
+    @product = product
+    @quantity = quantity
+  end
+
+  def add_to_cart
+    validate_quantity!
+
+    cart_product = CartProduct.find_or_initialize_by(cart: @cart, product: @product)
+    cart_product.quantity += @quantity
+    cart_product.total_amount = @product.price * cart_product.quantity
+    cart_product.save!
+
+    update_cart
+  end
+
+  def remove_from_cart
+    validate_quantity!
+
+    cart_product = CartProduct.find_by(cart: @cart, product: @product)
+    raise Services::ProductToRemoveNotAddedError unless cart_product
+
+    cart_product.quantity -= @quantity
+    if cart_product.quantity.negative?
+      error_data = { product_id: @product.id, quantity: @quantity }
+      raise Services::RemoveMoreProductsThanWereAddedError, error_data
+    end
+
+    product_extraction(cart_product)
+
+    update_cart
+  end
+
+  private
+
+  def update_cart
```

Comment:
```suggestion
  def update_cart!
```
you should use the bang to indicate that the method is safe

Reply:
Updated.